File: 14-longest-common-prefix/14-longest-common-prefix.py
from functools import reduce

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        output = ""
        max_length = len(reduce(lambda x, y : x if len(x) > len(y) else y, strs))
        i = 0
        
        while i < max_length:            
            for j in range(len(strs)-1):
                if i == len(strs[j]) or i == len(strs[j+1]):
                    return output
                
                if strs[j][i] != strs[j+1][i]:
                    return output
            
            output += strs[0][i]
            i += 1
        
        return output
        
        

File: 14-longest-common-prefix/14-longest-common-prefix.java
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if (strs == null || strs.length == 0)
            return "";        
        
        int minLen = Integer.MAX_VALUE;
        for (String str: strs) {
            minLen = Math.min(str.length(), minLen);
        }
        
        int low = 1;
        int high = minLen;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (isCommonPrefix(strs, mid))
                low = mid + 1;
            else
                high = mid - 1;
        }
        
        int mid = ((low + high) / 2);
        return strs[0].substring(0, mid);
    }
    
    private boolean isCommonPrefix(String[] strs, int len) {
        String substr = strs[0].substring(0, len);
        for (int i = 1; i < strs.length; i++)
            if (!strs[i].startsWith(substr))
                return false;        
        return true;
    }
}